spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java

```diff
@@ -0,0 +1,310 @@
+ */
+public class BigQueryFileMessageHandler extends AbstractReplyProducingMessageHandler {
+
+	private static final String HEADER_FORMAT = "headers[%s]";
+
+	private final BigQuery bigQuery;
+
+	private EvaluationContext evaluationContext;
+
+	private Expression datasetNameExpression;
+
+	private Expression tableNameExpression;
+
+	private Expression formatOptionsExpression;
+
+	private boolean autoDetectSchema = true;
+
+	private WriteDisposition writeDisposition = WriteDisposition.WRITE_APPEND;
+
+	private Duration timeout = Duration.ofMinutes(5);
+
+	private Duration jobPollInterval = Duration.ofSeconds(2);
+
+	private boolean sync = false;
+
+	public BigQueryFileMessageHandler(BigQuery bigQuery) {
+		Assert.notNull(bigQuery, "BigQuery client must not be null.");
+		this.bigQuery = bigQuery;
+
+		ExpressionParser expressionParser = new SpelExpressionParser();
+		this.datasetNameExpression = expressionParser.parseExpression(
+				String.format(HEADER_FORMAT, BigQuerySpringMessageHeaders.DATASET_NAME));
+		this.tableNameExpression = expressionParser.parseExpression(
+				String.format(HEADER_FORMAT, BigQuerySpringMessageHeaders.TABLE_NAME));
+		this.formatOptionsExpression = expressionParser.parseExpression(
+				String.format(HEADER_FORMAT, BigQuerySpringMessageHeaders.FORMAT_OPTIONS));
+	}
+
+	@Override
+	protected void doInit() {
+		this.evaluationContext = ExpressionUtils.createStandardEvaluationContext(getBeanFactory());
+	}
+
+	/**
+	 * Sets the SpEL {@link Expression} to evaluate to determine the dataset name.
+	 * @param datasetNameExpression the SpEL expression used to evaluate the dataset name
+	 */
+	public void setDatasetNameExpression(Expression datasetNameExpression) {
+		Assert.notNull(datasetNameExpression, "Dataset name expression can't be null.");
+		this.datasetNameExpression = datasetNameExpression;
+	}
+
+	/**
+	 * Sets the BigQuery dataset name to use.
+	 * @param datasetName name of the BigQuery dataset
+	 */
+	public void setDatasetName(String datasetName) {
+		this.datasetNameExpression = new LiteralExpression(datasetName);
+	}
+
+	/**
+	 * Sets the SpEL {@link Expression} to evaluate to determine the table name.
+	 * @param tableNameExpression the SpEL expression used to evaluate the table name
+	 */
+	public void setTableNameExpression(Expression tableNameExpression) {
+		this.tableNameExpression = tableNameExpression;
+	}
+
+	/**
+	 * Sets the BigQuery table name to use.
+	 * @param tableName name of the BigQuery table
+	 */
+	public void setTableName(String tableName) {
+		this.tableNameExpression = new LiteralExpression(tableName);
+	}
+
+	/**
+	 * Sets the default {@link FormatOptions} to use for the handler.
+	 * The {@link FormatOptions} describe the type/format of data files being loaded.
+	 * @param formatOptions the format of the data file being loaded
+	 */
+	public void setFormatOptions(FormatOptions formatOptions) {
+		this.formatOptionsExpression = new FunctionExpression<Message>(m -> formatOptions);
```

Comment:
This must be a `ValueExpression` instead. No reason in an intermediate `Function`